numba/mviewbuf.c

```diff
@@ -1,5 +1,199 @@
+        PyErr_SetString(PyExc_BufferError, "Unsupported flags for BufferProxy");
+        view->obj = NULL;
+        return -1;
+    }
+
+    if (self->ob_exports == 0) {
+        if(-1 == PyObject_GetBuffer(self->wrapped, &(self->wrapped_buf),
+                                    wrapped_flags)) {
+            PyErr_SetString(PyExc_BufferError,
+                            "Could not get buffer for wrapped (new protocol)");
+            view->obj = NULL;
+            return -1;
+        }
+    }
+
+    if(-1 == PyBuffer_FillInfo(view, (PyObject*) self, self->wrapped_buf.buf,
+                               self->wrapped_buf.len, 0, supported_flags)) {
+        PyBuffer_Release(&(self->wrapped_buf));
+        view->obj = NULL;
+        return -1;
+    }
+
+    self->ob_exports += 1;
+    return 0;
+}
+
+static void
+BufferProxyObject_releasebuffer(BufferProxyObject *self, Py_buffer *view) {
+    self->ob_exports -= 1;
+    if (self->ob_exports == 0)
+        PyBuffer_Release(&(self->wrapped_buf));
+}
+
+#if (PY_MAJOR_VERSION >= 3)
+static PyBufferProcs BufferProxy_as_buffer = {
+    (getbufferproc)BufferProxyObject_getbuffer,
+    (releasebufferproc)BufferProxyObject_releasebuffer,
+};
+#else
+static Py_ssize_t
+BufferProxyObject_getwritebuf(BufferProxyObject *self, Py_ssize_t index,
+                              const void **ptr) {
+    if (index != 0) {
+        PyErr_SetString(PyExc_TypeError,
+                        "Accessing non-existent BufferProxy segment");
+        return -1;
+    }
+
+    if (self->wrapped == NULL) {
+        PyErr_SetString(PyExc_TypeError, "No wrapped object");
+        return -1;
+    }
+
+    if (self->ob_old_exports == 0) {
+        if (-1 == PyObject_AsReadBuffer(self->wrapped,
+                                        (const void**) &(self->wrapped_ptr),
+                                        &(self->wrapped_len))) {
+            PyErr_SetString(PyExc_TypeError,
+                            "Could not get buffer for wrapped (old protocol)");
+            return -1;
+        }
+    }
+
+    self->ob_old_exports += 1;
+    *ptr = self->wrapped_ptr;
+    return self->wrapped_len;
+}
+
+static Py_ssize_t
+BufferProxyObject_getreadbuf(BufferProxyObject *self, Py_ssize_t index,
+                             const void **ptr) {
+    return BufferProxyObject_getwritebuf(self, index, ptr);
+}
+
+static Py_ssize_t
+BufferProxyObject_getsegcount(BufferProxyObject *self, Py_ssize_t *lenp) {
+    return self->wrapped->ob_type->tp_as_buffer->bf_getsegcount(self->wrapped,
+                                                                lenp);
+}
+
+static Py_ssize_t
+BufferProxyObject_getcharbuf(BufferProxyObject *self, Py_ssize_t index,
+                             const char **ptr) {
+    return BufferProxyObject_getwritebuf(self, index, (const void **) ptr);
+}
+
+static PyBufferProcs BufferProxy_as_buffer = {
+    (readbufferproc)BufferProxyObject_getreadbuf,       /*bf_getreadbuffer*/
```

Comment:
Is it useful to implement the old buffer API? AFAIU, this type is only used internally.